test/should-compile/InstanceConstraints/Test.hs

```diff
@@ -42,7 +42,7 @@ f5 :: forall a b x y
 f5 = foo @(F a x)
 
 f6 :: forall a b x y
-    . ( F a b ~ G x y
+    . ( b ~ G x y
```

Comment:
This looks suspicious. If the test has to be changed in order for the test to pass, that means the test is failing, no?

Reply:
I see that in ghc-9.2, the unmodified test fails with

```
    • Could not deduce (Foo (F a (G x x))) arising from a use of ‘foo’
      from the context: (F a b ~ G x y, x ~ y, Foo (F a b))
        bound by the type signature for:
                   f6 :: forall a b x y.
                         (F a b ~ G x y, y ~ x, x ~ y, Foo (F a b)) =>
                         ()
        at test/should-compile/InstanceConstraints/Test.hs:(44,1)-(50,8)
    • In the expression: foo @(F a (G x y))
      In an equation for ‘f6’: f6 = foo @(F a (G x y))
```

Reply:
The commit which introduces this test says:

> test that substitutions also work with non-variables
> 
> the code only substitutes variables, but that's good enough because ghc
> replaces nested calls to type families with fsk type variables whose
> definition is in the type substitution.

The body of `f6` is `foo @(F a (G x y))`, which requires a `Foo (F a (G x y))` instance. The constraint `F a b ~ G x y` means that ghc _could_ decide to simplify the requirement to `Foo (F a (F a b))`. This would allow the rewrite rule `forall a b. F a (F a b) ~ F a b` to fire, thus simplifying the requirement to `Foo (F a b)`, which matches the given.

The difficulty is that ghc has no reason to choose to make that "simplification", because `F a b` is not in any way simpler than `G x y`. It is thus the job of typelevel-rewrite-rules to look at the available equality constraints and conclude that `Foo (F a (G x y))` can be simplified to `Foo (F a b)`.

In other words, this is a legitimate test. Changing the constraint from `F a b ~ G x y` to `b ~ G x y` completely changes the test:
1. `b` is simpler than `G x y`, so ghc is more likely to decide to make the simplification on its own, so the test no longer checks that typelevel-rewrite-rules is able to use the available equality constraints to look a few steps ahead.
2. Once ghc performs the simplification, we are left with `Foo (F a b)`, which already matches the given without requiring typelevel-rewrite-rules to rewrite anything.

Reply:
Looking at the contents of `typeSubst`, the problem is clear: ghc-9.2 no longer replaces nested calls to type families with type variables named `fskN`. typelevel-rewrite-rules thus needs to work harder to determine whether rewrite rules should fire.

Reply:
Thank you for your review.
The reason why hasktorch didn't generate an error was that this condition was not used by chance.

Reply:
I'll see how other projects deal with it.

Reply:
I can think of two equivalent solutions.

1. do what GHC was doing: replace every sub-expression with a variable. Once we determine that the rewrite can be performed, we must carefully undo those replacements so that we only send GHC a replacement which involves its type variables, not ours.
2. use the unification algorithm (which also allocates one unification variable per sub-expression!) to determine which expression is equal to what. Then, check if any of those type variables unify with the rewrite rule's input. One subtlety is that we can't fail the unification when we encounter mismatched head constructors, because those might be stuck type families, which are allowed to be equal to each other.

To determine which of those two approaches would be the easiest to implement, I need to remind myself how we're already using unification and how we are currently distinguishing between ghc's unification variables and ours.

Reply:
I think we might also need to add a test where the expression to be substituted is inside a larger expression. Previously, with GHC separating all the sub-expressions, it was enough to check if the RHS of a `fskN ~ F a b` constraint matched the rewrite pattern, but it is now necessary to look deeper inside both the LHS and the RHS of `F x (F a b) ~ G x (F a b)`.

Reply:
> I need to remind myself [...] how we are currently distinguishing between ghc's unification variables and ours

The rewrite code is using two different substitution types:

```haskell
type Subst = Map TyVar TypeTerm
type TypeSubst = [(TypeEq, TypeTerm)]
```

Beside the poor naming choice (both hold types, so the "Type" prefix doesn't clarify the difference between the two) and the accidental use of two different data structure, the main difference is the key type: `TyVar` is a type variable, while `TypeEq` can be any type... But the code assumes it's a type variable 🤦

The one which uses `TyVar` keys holds the replacements for the variables in the LHS of a rewrite rule, while the one which uses `TypeEq` keys holds the `fsk1 ~ G x y` constraints from GHC.

The fact that `TypeEq` allows arbitrary types as keys is what allows typelevel-rewrite-rules not to barf on ghc-9.2's `F a b ~ G x y` constraints: the key is now `F a b`. But that doesn't work because if the type expression to be rewritten contains an `F a b`, that part will be expanded to `Fun "F" [Var "a", Var "b"]` before the rewrite begins, and it is only at the leaves that the rewrite code checks whether `"a"` or `"b"` matches the key `F a b`. The substitution is thus not performed, and that's why the test fails on ghc-9.2.

Reply:
Taking a closer look, the data structure choice is not accidental at all! ghc simplifies the constraint `F a b ~ G x y` to _two_ constraints, `fsk1 ~ F a b` and `fsk1 ~ G x y`. Normally an association list with duplicate keys just wastes space because `Data.List.lookup` stops after the first occurrence, but the rewrite code looks at _all_ the possible substitutions. This has exponential cost, but has the benefit of being correct!

Reply:
Thinking about this further, there is no reason to expect that `fsk2 ~ H u v` should still hold. If the rewrite rules says that `F a (F a b)` should rewrite to `F a`, that's a claim about a property of `F`, it does not imply that the rewrite should be performed because `F a b ~ F a`. So I now think that the current code is correct, for pre-ghc-9.2 versions at least.